src/com/intellij/stats/ngram/NGramContributor.kt

```diff
@@ -0,0 +1,54 @@
+package com.intellij.stats.ngram
+
+import com.intellij.codeInsight.completion.CompletionContributor
+import com.intellij.codeInsight.completion.CompletionParameters
+import com.intellij.codeInsight.completion.CompletionResult
+import com.intellij.codeInsight.completion.CompletionResultSet
+import com.intellij.codeInsight.lookup.LookupElement
+import com.intellij.codeInsight.lookup.LookupElementWeigher
+import com.intellij.openapi.project.Project
+import com.intellij.psi.search.GlobalSearchScope
+
+class NGramContributor : CompletionContributor() {
+    override fun fillCompletionVariants(parameters: CompletionParameters, result: CompletionResultSet) {
+        result.runRemainingContributors(parameters, { completionResult ->
+            val wrapped = CompletionResult.wrap(completionResult.lookupElement,
+                    completionResult.prefixMatcher,
+                    completionResult.sorter.weigh(NaiveNGramWeigher(parameters)))
+            if (wrapped != null) {
+                result.passResult(wrapped)
+            }
+        })
+    }
+}
+
+abstract class NGramWeigher(parameters: CompletionParameters, name : String) : LookupElementWeigher(name) {
+    val nGram: NGram
+    val project: Project
+
+    init {
+        val originalPosition = parameters.position
+        nGram = NGram.getNGramForElement(originalPosition)
```

Comment:
Is it possible to perform this operation in O(1)? Seems, tree traversal on each completion could be too expensive for large files

Reply:
AFAIK on each completion, Idea makes copy the file with `IntelliJRulezzz` identifier, so I don't think that an additional traverse of the whole file on each completion is a big deal.